### DISCUSSION TITLE
Internal Proj Error: proj_create: no database context specified

### DISCUSSION THREAD

**Discussion by abashishvili** [pyproj4/pyproj#1262] [Q&A]
Hello Everyone, 

Its already been a long time since I have had this error when executing the code, and I was not able to find a solution on the internet. 

So the problem is the following, I created the new conda environment and installed geopandas. When I import geopandas, I get this warning: 

```
UserWarning: pyproj unable to set database path. _pyproj_global_context_initialize()
```

Whenever I want to do a simple operation, for example, import shape file and change crs, I get the following error:

```
Invalid projection: EPSG:4326: (Internal Proj Error: proj_create: no database context specified)
```

I searched for the problem, and it seems pyproj database is mixed, and that's causes the error. There is something mentioned on pyproj FAQ about the database directories, but it was difficult for me to fully understand how to solve it. Any suggestion? 

Thanks for your help!

**Comment by snowman2** [comment]
This is likely due to installation issues. Mind providing this information?

#### Installation method/steps
 - Installation method (conda, pip wheel, from source, etc...)
 - How did you install PROJ? Where is it installed?
 - Please provide all commands/steps you used to install pyproj and PROJ.

#### Environment Information
 - pyproj version you are attempting to install
 - PROJ version (Execute `proj` command and give version here.)
 - Python version (`python -c "import sys; print(sys.version.replace('\n', ' '))"`)
 - Operation System Information (`python -c "import platform; print(platform.platform())"`)

**Reply by abashishvili** [reply]
Hello, 

Thanks so much for your answer. So the reason I came across to this problem is when I try to use  [osmnx package](https://osmnx.readthedocs.io/en/stable/).

Here is the requisted information: 

**Installation method/steps**
I follow the [osmnx package](https://osmnx.readthedocs.io/en/stable/) installation guide and install `osmnx` using Anaconda promt: 
```
conda config --prepend channels conda-forge
conda create -n ox_test --strict-channel-priority osmnx
```
As you can see it creates the new environment and also installs all necessary packages (see attached text file), including `pyproj` and `PROJ`.

[packege_list.txt](https://github.com/pyproj4/pyproj/files/11109745/packege_list.txt)

**Environment Information**
`pyproj 3.5.0 `
`proj 9.1.1`
`3.11.0 | packaged by conda-forge | (main, Oct 25 2022, 06:12:32) [MSC v.1929 64 bit (AMD64)]`
`Windows-10-10.0.22621-SP0`

When I using this environment and run the code `import geopandas as gpd` . I get the warning `[c:\Users\Win7ADM\anaconda3\envs\ox_test\Lib\site-packages\pyproj\__init__.py:89](file:///C:/Users/Win7ADM/anaconda3/envs/ox_test/Lib/site-packages/pyproj/__init__.py:89): UserWarning: pyproj unable to set database path.
  _pyproj_global_context_initialize()`

Please let me know if you require additional information! 

Best,

**Reply by snowman2** [reply]
Here are some things to check:

1. Did you run the command `conda activate ox_test` before using geopandas?
2. What are the values of these environment variables: `PROJ_DATA` and `PROJ_LIB` ?

**Reply by abashishvili** [reply]
Thanks for your reply! 

1. Yes, I work in VScode and it gives you option to select the interpreter. I also double checked this by running `conda activate ox_test` in Anaconda prompt and then type `code`, which will open the VScode in `ox_test` environment. It produces the same error
2.  Is this the right way to get the values of the environment variables? (if yes, please see the text file attached, which lists all the values of environment variables)
```
import os
print(os.environ)
````
[env_var.txt](https://github.com/pyproj4/pyproj/files/11116140/env_var.txt)

If I am wrong in providing further details, please let me know! 

Best

**Reply by snowman2** [reply]
This looks correct:
```
'PROJ_DATA': 'C:\\Users\\Win7ADM\\anaconda3\\envs\\ox_test\\Library\\share\\proj'
```

**Reply by abashishvili** [reply]
Could be this problem related to the error mentioned in [FAQ on pyproj website? ](https://pyproj4.github.io/pyproj/stable/gotchas.html#internal-proj-error-sqlite-error-on-select)

**Reply by snowman2** [reply]
Not exactly, it is not finding a PROJ database. What is the output of: `pyproj.datadir.get_data_dir`?

**Reply by abashishvili** [reply]
It is `C:\Users\Win7ADM\anaconda3\envs\ox_test\Library\share\proj`

**Reply by snowman2** [reply]
What are the contents of that directory?

**Comment by avriiil** [comment]
I'm running into the same issue. 

I've installed via `mamba install pyproj`. 

Here are the contents of my `proj_dir/share/proj` directory:

```
CH				deformation_model.schema.json	proj.db
GL27				nad.lst				proj.ini
ITRF2000			nad27				projjson.schema.json
ITRF2008			nad83				triangulation.schema.json
ITRF2014			other.extra			world
```

**Reply by snowman2** [reply]
https://github.com/pyproj4/pyproj/discussions/1262#discussioncomment-5469017

**Reply by avriiil** [reply]
python version: 3.12
OS: MacOS 14.3.1
`proj`: Rel. 9.5.0, September 15th, 2024
PROJ not installed explicitly, just via `mamba install pyproj` AFAIK

trying to install latest pyproj

**Reply by snowman2** [reply]
Did you call `mamba activate` to activate your virtual environment?

**Comment by avriiil** [comment]
Yes

**Reply by snowman2** [reply]
Are all of the packages from conda-forge?

**Comment by sheilasaia** [comment]
Hi all, I recently ran into this same issue trying to run on Windows in VS Code. I found that when I activated my environment in the Anaconda Powershell Prompt using `conda activate myenv` and then checked the environment using `dir env:`, my PROJ_DATA and PROJ_LIB paths were not pointing to the proj directory within the environment. The correct path should look something like `C:Users/yourname/AppData/Local/anaconda3/envs/myenv/Library/shared/proj` for both the PROJ_DATA and PROJ_LIB variables. If you're also finding that the paths are not pointing to the correct place, you can fix them permanently within your environment using `conda env config vars set PROJ_DATA =C:Users/yourname/AppData/Local/anaconda3/envs/myenv/Library/shared/proj`. You can also set the PROJ_LIB variable as well, but I think the PROJ_DATA is more important for newer versions of pyproj. You will then need to deactivate your conda environment (use `conda deactivate`) and then reactivate using `conda activate myenv` again for the changes to the environmental variables to save. Then you should be able to open VS code from this terminal window using `code` and it will automatically be set to your environment and use the environmental variables you just set. If you want to make 100% sure it's set up ok, you can import pyproj (`import pyproj`) and then run `pyproj.show_versions()`. There is a similar show versions command for rasterio too.

I also found that there was something weird going on with the order in which you import pyproj and rasterio, but it sounds like that's still an open issue [here](https://github.com/conda-forge/rasterio-feedstock/issues/250). TLDR; an ad-hoc fix for this is to `import osgeo` before you import either pyproj or rasterio.

I hope that helps!
`

**Reply by djhoese** [reply]
This discussion is very old and many things have likely changed. It is important to include what versions of the software you are installing and from where. These environment variables are set in the PROJ (`proj`) package from conda-forge and are configured as part of the activate scripts here:

https://github.com/conda-forge/proj.4-feedstock/tree/main/recipe/scripts

If these settings in these scripts are not correct or not complete then please file an issue there so it can get fixed.

**Comment by sheilasaia** [comment]
@djhoese, thanks very much for sharing that link. I hadn't come across this resource. It's also good to know that PROJ_DATA is the variable for setting the PROJ location now. I'm similar to the original poster here in that I'm fairly new to using python geopspatial packages (and conda). I've found that the documentation can sometimes be difficult to decipher so I wanted to post what worked for me in case someone came across this GitHub page in a similar manner as I had.

Are these scripts in the repo you shared typically installed automatically when PROJ is installed with conda? If not, where do you recommend implementing the scripts in the link you provided or if are they designed to sit in a directory in my environment and run when the `conda activate <your env>`? Do you have any recommended pyproj or other documentation on this set up/where they are typically added? I'm looking [here](https://proj-tmp.readthedocs.io/en/latest/index.html), but don't see anything about these files. Thanks very much for your help and thanks for your patience with my questions! If there's a better place to post these than on this GitHub thread, please let me know.

**Reply by djhoese** [reply]
So `conda-forge` is one channel that provides conda packages. I assumed that that is where your packages were being installed from. Anaconda (the curated selection of conda packages) may have a different version and likely older than what is available from conda-forge due to the amount of testing and curating they do. I highly recommend using conda-forge packages since they are much more up to date.

The activation scripts should installed as part of the package installation and are meant to be automatically executed when you load (activate) your environment. You shouldn't need to do anything. If you're running an older version of these packages though, these scripts may have not been implemented or properly configured.

As for the documentation you're looking at, that URL is for `proj-tmp` which...I have no idea what that is. You may want:

https://proj.readthedocs.io/en/latest/index.html

It is not guaranteed that a software's official documentation will mention these activation scripts as:

1. conda package building is a completely separate process and repository outside of the main package distribution.
2. Users usually don't need to know or worry about these files.

Bottom line: With modern versions of packages this is just meant to work.

**Reply by sheilasaia** [reply]
Thanks very much, @djhoese!
